### PR TITLE
feat: add direct PTX fused-kernel blueprint

### LINKED ISSUE #865: [PTX] Complete and validate the direct-PTX runtime and fused-kernel blueprint foundation
Parent epic: #833

This issue owns the foundation and golden-example work in PR #832. It establishes the single repeatable blueprint that every later PTX-only kernel issue must follow; it does not replace or close the epic.

## Scope

- Finish the CUDA Driver API/PTX runtime, feature gates, deterministic fallback, cache/autotune behavior, capture compatibility, disposal, and diagnostics.
- Define and enforce the contiguous-layout/stride ABI used to admit direct memory-to-register kernels without dynamic stride checks in hot code.
- Provide representative fused online-attention and residual/RMSNorm kernels as executable golden examples.
- Provide the benchmark, correctness, allocation, and Nsight Compute evidence assembly line reused by child kernel PRs.

## Acceptance criteria

- [ ] Exactly one implementation PR targets `main`, contains `Closes #<this issue>`, and remains draft until every criterion below has evidence.
- [ ] Supported shapes/layouts use an explicit contiguous-layout contract; views, offsets, unsupported strides, dtypes, architectures, and edge sizes are rejected before launch and follow the documented fallback path.
- [ ] PTX loading, launch, caching, autotune, feature gating, CUDA graph/capture behavior, failure diagnostics, and resource disposal are covered by automated tests.
- [ ] Fused online attention performs tiled/online softmax without materializing the score matrix; representative residual/RMSNorm fusion avoids intermediate global-memory round trips.
- [ ] Correctness is demonstrated across the documented shape/dtype/mask/causal edge-case matrix against an independent reference, with maximum absolute/relative error and tolerances reported.
- [ ] Apples-to-apples NVIDIA GPU benchmarks compare the new PTX path with current AiDotNet GPU behavior, PyTorch CUDA/SDPA, and applicable vendor-library-backed GPU baselines under identical shapes, dtypes, warmup, synchronization, and timing boundaries.
- [ ] Every reported scenario includes median, P95, P99, mean, throughput (TFLOP/s or GFLOP/s, plus GB/s where appropriate), managed allocation, temporary VRAM, and numerical error, with the complete environment and commands recorded.
- [ ] Results come from at least three independent clean processes, reject concurrent GPU workloads/thermal-invalid captures, and show at least 1.10x speedup over the strongest applicable competitor with no P95 regression for the declared winning shapes.
- [ ] Nsight Compute executed-counter evidence reports registers/thread, static/dynamic shared memory, occupancy, local-memory traffic, and register-spill counters; all admitted golden-example kernels show zero executed spills and zero unintended local-memory loads/stores.
- [ ] The admitted PTX path performs zero managed allocations per steady-state call and zero temporary VRAM allocation per steady-state launch, with allocation ownership documented.
- [ ] Release builds and focused tests pass for every supported target framework; CI, patch coverage, and actionable review findings are green/resolved.
- [ ] The blueprint documentation is sufficient for another developer to implement a later PTX-only child kernel without inventing a new workflow, evidence format, or admission contract.

## Non-goals

- CPU MKL/OpenBLAS comparisons are excluded from the GPU championship table because they are not apples-to-apples NVIDIA GPU competitors.
- Completion of every operation tracked by the epic remains in the operation-specific child issues.

### PR DESCRIPTION
**Pull request by ooples**
Closes #865

## Summary  Introduces an opt-in, CUDA Driver API + runtime-emitted PTX blueprint for shape-specialized fused NVIDIA kernels without CUDA Runtime or NVRTC in the custom path.  The PR proves the process with:  - async tiled online attention for FP16 BHSD, D=64, S in {16,32,64,128} - unmasked and causal variants - optional fused FP32 LayerNorm + tanh-GELU epilogue - a second independent fusion: FP32 `RMSNorm(input + residual)`, D=64 - formal layout/extent/alignment/access ABI validation at the dispatch boundary - per-GPU/driver architecture identity, resource manifests, bounded module caches, autotuning, prewarm, and graph-capture-safe dispatch - explicit eligibility/fallback reasons for unsupported semantics - correctness, benchmark, release-gate, CUDA-event, external PyTorch, and Nsight Compute profiler harnesses  ## Why  The existing CUDA/NVRTC and library-composition paths materialize intermediates and repeat generic dispatch work. The new blueprint validates tensor layout once, emits stride-free PTX for a proven specialization, pipelines global memory through shared memory/registers, uses Tensor Core `mma.sync`, and fuses epilogues before the single final global write.  This PR is the foundation for a repository-wide PTX fused-kernel migration. It does not advertise unsupported shapes or architectures: unproven cells fail closed to the existing backend.  ## Measured evidence  RTX 3080 / SM86:  - unmasked direct attention: 33.69 us device median, 15.936 effective TFLOPS, 0 B managed allocation, 0 temporary VRAM - causal direct attention: 25.60 us device median, 20.972 effective TFLOPS, 0 B managed allocation, 0 temporary VRAM - cuBLAS composition: 51.30/51.51 us median and 12 MiB temporary VRAM - residual + RMSNorm median speedup over current AiDotNet: 1.15x to 2.04x across tested row counts - Driver JIT audit: attention 66-80 registers/thread, 25,088 B shared, 0 local B/thread; residual RMSNorm 18 registers/thread, 0 shared, 0 local B/thread  CUDA PyTorch 2.12.1+cu130 was installed and the forced cuDNN/Efficient/Math external matrix was executed. PyTorch Flash-SDPA and the third-party `flash-attn` package are unavailable in this Windows environment and are reported as explicit skips.  Nsight Compute 2025.4.1 attaches to the deterministic target, but the driver returns `ERR_NVGPUCTRPERM`. Executed hardware-counter proof therefore remains intentionally unclaimed until an administrator enables counter access.  ## Validation  - `dotnet build src/AiDotNet.Tensors/AiDotNet.Tensors.csproj -c Release --no-restore -p:UseSharedCompilation=false`   - net471, net8.0, net10.0: 0 warnings, 0 errors - `dotnet test tests/AiDotNet.Tensors.Tests/AiDotNet.Tensors.Tests.csproj -c Release -f net10.0 --filter DirectPtxWmmaTests --no-restore -p:UseSharedCompilation=false`   - 47 passed, 0 failed, 0 skipped - benchmark project Release/net10.0 build: succeeded; existing unrelated warnings remain - external forced CUDA PyTorch benchmark: succeeded for cuDNN/Efficient/Math - Nsight deterministic target: attached; counter collection blocked by host permission as described above  ## Safety and rollout  - disabled by default - opt-in with `AIDOTNET_DIRECT_PTX=1` or per-kernel feature gates - unsupported dtype/layout/shape/mask/training/architecture combinations fall back - architecture families are separate dispatch domains; this PR admits only the measured Ampere implementation - production release policy requires >=1.10x median, bounded P95, zero hot allocations/temporary VRAM, numerical tolerance, zero local bytes, and three clean independent runs  The complete benchmark matrix, ABI, dataflow, acceptance policy, commands, and remaining evidence boundaries are in `docs/research/2026-07-20-fused-attention-championship-blueprint.md`. ## Tracking  - PTX migration epic: #833 - Full attention-family follow-up: #834 - Normalization/residual-family follow-up: #838  This foundation PR proves representative attention and residual+RMSNorm specializations; it does not close the broader operation-family children.    ## Summary by CodeRabbit  - **New Features**   - Added an opt-in Direct PTX CUDA acceleration path for supported FP16 attention workloads on compatible NVIDIA GPUs.   - Added fused residual addition and RMSNorm acceleration for 64-wide vectors.   - Added optional fused LayerNorm and GELU processing for eligible attention operations.   - Added automatic kernel selection, caching, prewarming, and safe fallback to existing GPU implementations.  - **Documentation**   - Added research, benchmarking, profiling, eligibility, and reproduction guidance for the new GPU paths.  - **Tests**   - Added extensive correctness, compatibility, CUDA graph, and performance validation coverage.

### CODE REVIEW

**Comment by ooples** [comment]
@coderabbitai review

File: tests/AiDotNet.Tensors.Benchmarks/DirectPtxExternalBaselines.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+            throw new FileNotFoundException("The direct-PTX Python competitor harness was not copied to output.", script);
+
+        var start = new ProcessStartInfo
+        {
+            FileName = Environment.GetEnvironmentVariable("PYTHON") ?? "python",
+            UseShellExecute = false,
+            RedirectStandardOutput = false,
+            RedirectStandardError = false
+        };
+        start.ArgumentList.Add(script);
```
[the remark this answers is not in the document]
Comment (ooples) [reply tests/AiDotNet.Tensors.Benchmarks/DirectPtxExternalBaselines.cs:23]: Not a valid TFM issue — dismissing. `AiDotNet.Tensors.Benchmarks` is single-target `<TargetFramework>net10.0</TargetFramework>`; net471 is not a target for this project, so `ProcessStartInfo.ArgumentList` is available and correct. It is already used across 3 files in this project. Applying the proposed `.Arguments = \"…\"` change would replace the quoting-safe API with manual quoting for no benefit, so it is deliberately left as-is.

File: tests/AiDotNet.Tensors.Benchmarks/DirectPtxResidualRmsNormExperiment.cs [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+/// <summary>Second-blueprint NVIDIA-only fused residual + RMSNorm benchmark.</summary>
+internal static class DirectPtxResidualRmsNormExperiment
+{
+    private const int Dimension = 64;
+    private const float Epsilon = 1e-5f;
+    private readonly record struct Distribution(double Mean, double Median, double P95, double P99);
+    private readonly record struct Result(
+        int Rows, string Method, Distribution Time, double GigabytesPerSecond,
+        long Allocation, long TemporaryBytes, float MaxError, int Registers, int LocalBytes);
```
[the remark this answers is not in the document]
Comment (ooples) [reply tests/AiDotNet.Tensors.Benchmarks/DirectPtxResidualRmsNormExperiment.cs:17]: Not a valid TFM issue — dismissing. This project is single-target net10.0 (no net471), so `record struct` compiles fine and is already used in 9 files across the Benchmarks project. Guarding or hand-expanding it here would be inconsistent with the rest of the project for no compatibility gain.

File: tests/AiDotNet.Tensors.Benchmarks/DirectPtxResidualRmsNormExperiment.cs [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+    private static long Allocation(Action synchronize, Action launch)
+    {
+        launch(); synchronize();
+        long before = GC.GetAllocatedBytesForCurrentThread();
+        for (int i = 0; i < 50; i++) launch();
+        long allocation = (GC.GetAllocatedBytesForCurrentThread() - before) / 50;
+        synchronize();
+        return allocation;
+    }
```
[the remark this answers is not in the document]
Comment (ooples) [reply tests/AiDotNet.Tensors.Benchmarks/DirectPtxResidualRmsNormExperiment.cs:189]: Not a valid TFM issue — dismissing. Single-target net10.0 project; `GC.GetAllocatedBytesForCurrentThread()` is available and already used in 8 files across the Benchmarks project. No net471 target exists to guard for.
